src/main/java/org/rulez/demokracia/pdengine/CastVote.java

```diff
@@ -0,0 +1,26 @@
+package org.rulez.demokracia.pdengine;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import org.rulez.demokracia.pdengine.dataobjects.CastVoteEntity;
+
+
+public class CastVote extends CastVoteEntity {
+	private static final long serialVersionUID = 1L;
+
+	public CastVote(final String proxyId, final List<RankedChoice> preferences) {
+		super();
+		this.proxyId = proxyId;
+		this.preferences = new ArrayList<>(preferences);
+		secretId = RandomUtils.createRandomKey();
+	}
+
+	public List<RankedChoice> getPreferences() {
```

Comment:
> Using public fields for entities is not a bad habit, but a design choice.
> "The difference between a data structure and an object is that data structures have visible data and no behavior. Objects have behavior and no visible data."
> See https://sites.google.com/site/unclebobconsultingllc/active-record-vs-objects
> Of course I am open to discussion about either whether it is a legit approach or whether our way to implement the approach should be changed.

"What I am advocating is a separation between the application and Active Record."

Since we inherit objects with business logic from it, we can't handle VoteEntity as data structure.
In general, I prefer Beans over "Data structures" in OO design, because everything should be an object. I also think that null and primitives were mistakes in the design of Java language. I'm choosy :)
If we focus on it from the perspective of testing, in rare cases when you need to be sure about a field has been updated, you can't check it with a public field, but a spied setter method.

Reply:
If we actually take the "prefer composition over inheritance" rule seriously, then we only inherit maybe other entities from an entity. We implement model level functionality with traits (sets of related functionalities) as visitors of the entity. And here come some  questions:
1. There is clearly a need for a line of defence, for example to make sure that invariants are not violated. Getters and setters are obvious places to implement them.
2. Sometimes you need some kind of "inheritance" between traits. For example a "Voteable" trait (something on which votes can be cast) needs to operate on a "HaveChoices" object (something which offers choices). How it is best to implement it?
3. How can we figure out what kind of traits are applicable to something? Like that a Vote HaveChoices and Voteable.
4. We always run to the limitations that Java have single dependency and no fields can be defined in an interface.

Reply:
Solution
An entity (e.g. VoteEntity) is an object with public fields, and nothing else.
An object (e.g. Vote) is a visitor to the entity (VoteEntity), implementing only getters, setters and constructors, responsible for invariants and nothing else.
A trait (e.g. HaveChoices) is an interface implementing related  behaviours in default methods. 
The traits can extend other traits when needed, and the object implements all its traits, without having a single line of code for the traits.
Does it seem workable?

Reply:
According to my experience, the best practice is to separate data and business logic.
With a 3-layer architecture (Controller - Service - Repository) it's easy to implement.
All the classes of these layers are stateless (don't have any data field). They just define operations on data objects (entities and DTOs).
Controller calls only services, services may call each-other and the repository. Repository calls the DB.
If you use Spring framework, it  solves the dependency injection, so you don't need to take care about life-cycle of these components. 
1. Agree.
2. It sounds like a good candidate for the Decorator pattern. 
3. Composition of the 2 "base classes" will solve the problem you will need forward methods that calls the matching methods of the member objects. https://en.wikipedia.org/wiki/Composition_over_inheritance
4, BTW, multiple inheritance often violates the single responsibility principle, therefore it signs design problems.

Reply:
One more idea:
We store a bunch of boolean in the VoteEntity, Basically they should be different sub-classes.
E.g. 
public class VoteWithChoices extends Vote implements HaveChoices {
    private List<Choice> choices

    @Override
    public getChoices() {
        return this.choices;
    }
}

public interface HaveChoices {
    List<Choice> getChoices();

    default void doSomethingWithChoice(...) {
        List<Choice>.choicesWewillWorkOn = getChoices();
    ...
    }
}

So you define the behavior in a default method and expect a getter for all the needed fields.